### PR TITLE
Add Google sign-in

### PR DESCRIPTION
**Pull request by dvejmz**
This PR will add fully functional and secure Google-powered sign-in to the app. The only validation check missing is verifying Google Auth API's public key.

I've also done some housecleaning tasks like fixing typos and minor improvements to the Makefiles and .gitignore.

Don't be put off by the size of the PR, it's actually smaller than it looks. I added a new crate (`regex`) so I guess `cargo` went a bit nuts when updating `Cargo.lock`. I also split `AuthController` into `FacebookAuthController` and `GoogleAuthController`, the latter's basic structure has pretty much the same DNA footprint as the former.

Google auth first sign-in stage is done in a separate JavaScript file client-side whereas the second-stage (token verification) is done in the backend.

### CODE REVIEW

**Comment by dvejmz** [comment]
I've reset the `Cargo.lock` file to be the same as upstream because there were odd compilation errors otherwise. Is it really necessary that we keep `Cargo.lock` in source control? I know it's recommended for binaries and stuff but this wouldn't have happened if we just let cargo rebuild the file every time we get a brand new copy of the repo.

**Comment by benbrunton** [comment]
Cheer for this! Hoping to review tonight or at least by the weekend.

The reason there were conflicts is because I updated a couple of the dependencies. It's not a hard problem to solve and provides a bit of additional information about when there have been good builds so worth keeping as is imo.

http://doc.crates.io/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

File: project/client/js/google-auth.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+var googleUser = {};
+var initGoogleAuth = function() {
+    gapi.load('auth2', function(){
+        // Retrieve the singleton for the GoogleAuth library and set up the client.
+        auth2 = gapi.auth2.init({
+            client_id: '467471910617-vo4g6ojobn4o2tu25vo38ofurfis0jqj.apps.googleusercontent.com',
```
Comment (benbrunton) [inline project/client/js/google-auth.js:6]: What's this id? Does it need generating for play.benbru.com and should it be served to the frontend via some config?

Comment (dvejmz) [reply project/client/js/google-auth.js:6]: That is the client ID uniquely identifying the app to Google's API. In order to make Google OAuth possible, you'll need to generate a pair of unique IDs for play.benbru.com: one of them will be this, which can be shared publicly, and the other one is the secret ID which can't be made public. Instructions on how to do obtain these IDs can be found here: https://developers.google.com/identity/sign-in/web/devconsole-project

Once you've obtained those, add them to `app_config.tmpl` as with Facebook's:
```
# google credentials
google_secret = <SECRET>
google_app_id = <APP_ID>
```
I tried using Tera to fill in the ID straight from the config file by passing it as a parameter to the `initGoogleAuth()` function but this required those keys to be fed to every controller in order to support logout on any page.

Comment (benbrunton) [reply project/client/js/google-auth.js:6]: I guess for now I can just update with one I obtain myself and make a note to tackle this when I get a build pipeline sorted.

**Review by benbrunton — APPROVED** [review]

**Pull request merged**